src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs

```diff
@@ -426,26 +429,44 @@ internal static PropertyInfo[] GetKeyPropertiesOnType(Type type)
         /// <param name="type">Type in question.</param>
         /// <param name="hasProperties">true if <paramref name="type"/> has any (declared or inherited) properties; otherwise false.</param>
         /// <returns>Returns the list of key properties defined on <paramref name="type"/>; null if <paramref name="type"/> is complex.</returns>
-        internal static PropertyInfo[] GetKeyPropertiesOnType(Type type, out bool hasProperties)
+        internal static PropertyInfo[] GetKeyPropertiesOnType(Type type,DataServiceContext context,out bool hasProperties)
         {
             if (CommonUtil.IsUnsupportedType(type))
             {
                 throw new InvalidOperationException(c.Strings.ClientType_UnsupportedType(type));
             }
-
             string typeName = type.ToString();
             IEnumerable<object> customAttributes = type.GetCustomAttributes(true);
+            bool isKeyDeclared = false;
+            if (customAttributes.Count() == 0)
+            {
+                if (context != null)
+                {
+                    IEdmModel model = context.Format.ServiceModel;
+                    if (model != null && model is ClientEdmModel)
+                    {
+                        EdmEntityType edmEntityType = (EdmEntityType)model.SchemaElements.FirstOrDefault();
+                        if (edmEntityType != null)
+                        {
+                            var declaredKey = edmEntityType.DeclaredKey;
+                            if (declaredKey != null)

```

Comment:
These tests can be simplified with to reduce the nesting maybe having
```if (context?.Format.ServiceModel is ClientEdmModel model)``` maybe followed by ``` model.SchemaElements.FirstOrDefault()?.DeclaredKey...```

Reply:
Fixed